

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Parsers &mdash; GwasQcPipeline 1.0.0b2 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Workflow Scripts" href="workflow.html" />
    <link rel="prev" title="Configuration" href="config.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home" alt="Documentation Home"> GwasQcPipeline
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Getting Started:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/installation.html">Installing GwasQcPipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/running_pipeline.html">Running the Pipeline</a></li>
</ul>
<p class="caption"><span class="caption-text">QC Modules:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../qc_modules/sample_level.html">Sample QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../qc_modules/variant_level.html">Variant QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../qc_modules/reporting.html">QC Report</a></li>
</ul>
<p class="caption"><span class="caption-text">Module Reference:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../reference/cli.html">Command Line Interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/file_types.html">File Type Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/rules.html">Rule Reference</a></li>
</ul>
<p class="caption"><span class="caption-text">Developer Documentation:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../dev_docs/building_docs.html">Building Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dev_docs/dev_tools.html">Development Tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dev_docs/running_tests.html">Running Tests</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dev_docs/setup.html">Setting-up the Development Environment</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dev_docs/writing_docs.html">Writing Documentation</a></li>
</ul>
<p class="caption"><span class="caption-text">API Reference:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="cgr_gwas_qc.html">API Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="cli.html">Command Line Interface</a></li>
<li class="toctree-l2"><a class="reference internal" href="config.html">Configuration</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Parsers</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#module-cgr_gwas_qc.parsers.eigensoft">Eigensoft</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-cgr_gwas_qc.parsers.graf">GRAF</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genetic-relationship-and-fingerprinting">Genetic Relationship and Fingerprinting</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-cgr_gwas_qc.parsers.king">KING</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#king-relationship-inferences">KING Relationship Inferences</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-cgr_gwas_qc.parsers.plink">Plink</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="workflow.html">Workflow Scripts</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">GwasQcPipeline</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="cgr_gwas_qc.html">API Reference</a> &raquo;</li>
        
      <li>Parsers</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/api/parsers.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="parsers">
<h1>Parsers<a class="headerlink" href="#parsers" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-cgr_gwas_qc.parsers.eigensoft">
<span id="eigensoft"></span><h2>Eigensoft<a class="headerlink" href="#module-cgr_gwas_qc.parsers.eigensoft" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt id="cgr_gwas_qc.parsers.eigensoft.Eigenvec">
<em class="property">class </em><code class="sig-prename descclassname">cgr_gwas_qc.parsers.eigensoft.</code><code class="sig-name descname">Eigenvec</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">filename</span><span class="p">:</span> <span class="n">pathlib.Path</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/cgr_gwas_qc/parsers/eigensoft.html#Eigenvec"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cgr_gwas_qc.parsers.eigensoft.Eigenvec" title="Permalink to this definition">¶</a></dt>
<dd><p>Eigensoft eigenvec file parser.</p>
<dl class="py attribute">
<dt id="cgr_gwas_qc.parsers.eigensoft.Eigenvec.filename">
<code class="sig-name descname">filename</code><a class="headerlink" href="#cgr_gwas_qc.parsers.eigensoft.Eigenvec.filename" title="Permalink to this definition">¶</a></dt>
<dd><p>Path to the eigenvec file from <code class="docutils literal notranslate"><span class="pre">smartpca</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>Path</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="cgr_gwas_qc.parsers.eigensoft.Eigenvec.components">
<code class="sig-name descname">components</code><a class="headerlink" href="#cgr_gwas_qc.parsers.eigensoft.Eigenvec.components" title="Permalink to this definition">¶</a></dt>
<dd><p>A (n x 11) table of principal components</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>name</p></th>
<th class="head"><p>dtype</p></th>
<th class="head"><p>description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>ID</strong> (<em>index</em>)</p></td>
<td><p>object</p></td>
<td><p>Sample or Subject ID</p></td>
</tr>
<tr class="row-odd"><td><p>PC1</p></td>
<td><p>object</p></td>
<td><p>The first principal componenet (PC1).</p></td>
</tr>
<tr class="row-even"><td><p>PC2</p></td>
<td><p>object</p></td>
<td><p>The second principal componenet.</p></td>
</tr>
<tr class="row-odd"><td><p>…</p></td>
<td><p>…</p></td>
<td><p>The third to ninth principal components.</p></td>
</tr>
<tr class="row-even"><td><p>PC10</p></td>
<td><p>object</p></td>
<td><p>The tenth principal componenet.</p></td>
</tr>
</tbody>
</table>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="cgr_gwas_qc.parsers.eigensoft.Eigenvec.values">
<code class="sig-name descname">values</code><a class="headerlink" href="#cgr_gwas_qc.parsers.eigensoft.Eigenvec.values" title="Permalink to this definition">¶</a></dt>
<dd><p>A vector of eigenvalues for PC1 to PC10.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<p class="rubric">References</p>
<ul class="simple">
<li><p><a class="reference external" href="https://www.hsph.harvard.edu/alkes-price/software/">https://www.hsph.harvard.edu/alkes-price/software/</a></p></li>
<li><p><a class="reference external" href="https://www.cell.com/ajhg/fulltext/S0002-9297(16)00003-3">https://www.cell.com/ajhg/fulltext/S0002-9297(16)00003-3</a></p></li>
</ul>
</dd></dl>

</div>
<div class="section" id="module-cgr_gwas_qc.parsers.graf">
<span id="graf"></span><h2>GRAF<a class="headerlink" href="#module-cgr_gwas_qc.parsers.graf" title="Permalink to this headline">¶</a></h2>
<div class="section" id="genetic-relationship-and-fingerprinting">
<h3>Genetic Relationship and Fingerprinting<a class="headerlink" href="#genetic-relationship-and-fingerprinting" title="Permalink to this headline">¶</a></h3>
<p>GRAF is a package that allows estmation or relatedness and ancestry.</p>
<div class="section" id="relatedness">
<h4>Relatedness<a class="headerlink" href="#relatedness" title="Permalink to this headline">¶</a></h4>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>name</p></th>
<th class="head"><p>dtype</p></th>
<th class="head"><p>description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>ID1</p></td>
<td><p>string</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>ID2</p></td>
<td><p>string</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>HG_match</p></td>
<td><p>int</p></td>
<td><p>number of SNPs with matched genotypes when only homozygous SNPs are counted</p></td>
</tr>
<tr class="row-odd"><td><p>HG_miss</p></td>
<td><p>int</p></td>
<td><p>number of SNPs with mismatched genotypes when only homozygous SNPs are counted</p></td>
</tr>
<tr class="row-even"><td><p>HGMR</p></td>
<td><p>float</p></td>
<td><p>Homozygous Genotype Mismatch Rate (%)</p></td>
</tr>
<tr class="row-odd"><td><p>AG_match</p></td>
<td><p>int</p></td>
<td><p>number of SNPs with matched genotypes when all SNPs are counted</p></td>
</tr>
<tr class="row-even"><td><p>AG_miss</p></td>
<td><p>int</p></td>
<td><p>number of SNPs with mismatched genotypes when all SNPs are counted</p></td>
</tr>
<tr class="row-odd"><td><p>AGMR</p></td>
<td><p>float</p></td>
<td><p>All Genotype Mismatch Rate (%)</p></td>
</tr>
<tr class="row-even"><td><p>relationship</p></td>
<td><p>string</p></td>
<td><p>relationship determined by sample genotypes.</p></td>
</tr>
<tr class="row-odd"><td><p>p_value</p></td>
<td><p>float</p></td>
<td><p>probability that the genetic relationship is NOT the predicted type</p></td>
</tr>
</tbody>
</table>
<div class="section" id="relationship-values">
<h5>Relationship Values<a class="headerlink" href="#relationship-values" title="Permalink to this headline">¶</a></h5>
<p>Categories are assigned by GRAF.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>name</p></th>
<th class="head"><p>description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>ID</p></td>
<td><p>duplicate or MZ twin</p></td>
</tr>
<tr class="row-odd"><td><p>PO</p></td>
<td><p>parent-offspring</p></td>
</tr>
<tr class="row-even"><td><p>FS</p></td>
<td><p>full sibling</p></td>
</tr>
<tr class="row-odd"><td><p>D2</p></td>
<td><p>2nd degree relative</p></td>
</tr>
<tr class="row-even"><td><p>D3</p></td>
<td><p>3rd degree relative</p></td>
</tr>
<tr class="row-odd"><td><p>UN</p></td>
<td><p>unrelated</p></td>
</tr>
</tbody>
</table>
<p class="rubric">References</p>
<ul class="simple">
<li><p><a class="reference external" href="https://github.com/ncbi/graf">https://github.com/ncbi/graf</a></p></li>
<li><p>Jin Y, Schäffer AA, Sherry ST, and Feolo M (2017). Quickly identifying
identical and closely related subjects in large databases using genotype
data. PLoS One. 12(6):e0179106.</p></li>
</ul>
<dl class="py function">
<dt id="cgr_gwas_qc.parsers.graf.read_relatedness">
<code class="sig-prename descclassname">cgr_gwas_qc.parsers.graf.</code><code class="sig-name descname">read_relatedness</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">filename</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>os.PathLike<span class="p">[</span>str<span class="p">]</span><span class="p">, </span>pathlib.Path<span class="p">]</span></span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="reference internal" href="../_modules/cgr_gwas_qc/parsers/graf.html#read_relatedness"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cgr_gwas_qc.parsers.graf.read_relatedness" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads the table generated by <code class="docutils literal notranslate"><span class="pre">graf</span> <span class="pre">--out</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><p>pd.DataFrame</p>
<ul class="simple">
<li><p>ID1</p></li>
<li><p>ID2</p></li>
<li><p>HG_match</p></li>
<li><p>HG_miss</p></li>
<li><p>HGMR</p></li>
<li><p>AG_match</p></li>
<li><p>AG_miss</p></li>
<li><p>AGMR</p></li>
<li><p>relationship {ID, PO, FS, D2, D3, UN}</p></li>
<li><p>p_value</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<ul class="simple">
<li><p><a class="reference external" href="https://github.com/ncbi/graf#output-files">https://github.com/ncbi/graf#output-files</a></p></li>
</ul>
</dd></dl>

</div>
</div>
</div>
</div>
<div class="section" id="module-cgr_gwas_qc.parsers.king">
<span id="king"></span><h2>KING<a class="headerlink" href="#module-cgr_gwas_qc.parsers.king" title="Permalink to this headline">¶</a></h2>
<div class="section" id="king-relationship-inferences">
<h3>KING Relationship Inferences<a class="headerlink" href="#king-relationship-inferences" title="Permalink to this headline">¶</a></h3>
<p>KING is a toolset to robustly identify different types of relatedness. Unlike
other methods, you should only remove markers that fail QC (i.e., don’t mess
with MAF or LD filters).</p>
<div class="section" id="kinship-table">
<h4>Kinship Table<a class="headerlink" href="#kinship-table" title="Permalink to this headline">¶</a></h4>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>name</p></th>
<th class="head"><p>dtype</p></th>
<th class="head"><p>description</p></th>
<th class="head"></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>ID1</p></td>
<td><p>string</p></td>
<td><p>Individual ID for the first individual of the pair</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>ID2</p></td>
<td><p>string</p></td>
<td><p>Individual ID for the second individual of the pair</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>N_SNP</p></td>
<td><p>UInt32</p></td>
<td><p>The number of SNPs that do not have missing genotypes in either of the individual</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>HetHet</p></td>
<td><p>float</p></td>
<td><p>Proportion of SNPs with double heterozygotes (e.g.</p></td>
<td><p>AG and AG)</p></td>
</tr>
<tr class="row-even"><td><p>IBS0</p></td>
<td><p>float</p></td>
<td><p>Proportion of SNPs with zero IBS (identical-by-state) (e.g.</p></td>
<td><p>AA and GG)</p></td>
</tr>
<tr class="row-odd"><td><p>HetConc</p></td>
<td><p>float</p></td>
<td><p>Heterozygous concordance</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>HomIBS0</p></td>
<td><p>float</p></td>
<td><p>Homozygous IBS0</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>Kinship</p></td>
<td><p>float</p></td>
<td><p>Estimated kinship coefficient from the SNP data</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>IBD1Seg</p></td>
<td><p>float</p></td>
<td><p>Total length of IBD1 segments divided by total length of all segments</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>IBD2Seg</p></td>
<td><p>float</p></td>
<td><p>Total length of IBD2 segments divided by total length of all segments</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>PropIBD</p></td>
<td><p>float</p></td>
<td><p>Proportion of genome with shared IBD (e.g.</p></td>
<td><p>IBD2Seg + IBD1Seg/2)</p></td>
</tr>
<tr class="row-odd"><td><p>relationship</p></td>
<td><p>string</p></td>
<td><p>The assigned relationship based on Kinship</p></td>
<td></td>
</tr>
</tbody>
</table>
<div class="section" id="id1">
<h5>Relationship Values<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h5>
<p>Categories were assigned based on Kinship ranges provided in the <a class="reference external" href="http://people.virginia.edu/~wc9c/KING/manual.html#WITHIN">KING manual</a>.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>name</p></th>
<th class="head"><p>description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>ID</p></td>
<td><p>duplicate or MZ twin</p></td>
</tr>
<tr class="row-odd"><td><p>D1</p></td>
<td><p>1st degree relative</p></td>
</tr>
<tr class="row-even"><td><p>D2</p></td>
<td><p>2nd degree relative</p></td>
</tr>
<tr class="row-odd"><td><p>D3</p></td>
<td><p>3rd degree relative</p></td>
</tr>
<tr class="row-even"><td><p>UN</p></td>
<td><p>unrelated</p></td>
</tr>
</tbody>
</table>
<p class="rubric">References</p>
<ul class="simple">
<li><p><a class="reference external" href="http://people.virginia.edu/~wc9c/KING/manual.html">http://people.virginia.edu/~wc9c/KING/manual.html</a></p></li>
<li><p>Manichaikul A, Mychaleckyj JC, Rich SS, Daly K, Sale M, Chen WM (2010)
Robust relationship inference in genome-wide association studies.
Bioinformatics 26(22):2867-2873</p></li>
</ul>
<dl class="py function">
<dt id="cgr_gwas_qc.parsers.king.read_related">
<code class="sig-prename descclassname">cgr_gwas_qc.parsers.king.</code><code class="sig-name descname">read_related</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">filename</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>os.PathLike<span class="p">[</span>str<span class="p">]</span><span class="p">, </span>pathlib.Path<span class="p">]</span></span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="reference internal" href="../_modules/cgr_gwas_qc/parsers/king.html#read_related"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cgr_gwas_qc.parsers.king.read_related" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads the table generated by <code class="docutils literal notranslate"><span class="pre">king</span> <span class="pre">--related</span></code>.</p>
<p>Reads the king file and re-assigns ID1/ID2 by sorting IDs
alphanumerically.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><p>pd.DataFrame</p>
<ul class="simple">
<li><p>ID1</p></li>
<li><p>ID2</p></li>
<li><p>N_SNP</p></li>
<li><p>HetHet</p></li>
<li><p>IBS0</p></li>
<li><p>HetConc</p></li>
<li><p>HomIBS0</p></li>
<li><p>Kinship</p></li>
<li><p>IBD1Seg</p></li>
<li><p>IBD2Seg</p></li>
<li><p>PropIBD</p></li>
<li><p>relationship {ID, PO, FS, D2, D3}</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<ul class="simple">
<li><p><a class="reference external" href="http://people.virginia.edu/~wc9c/KING/manual.html#WITHIN">http://people.virginia.edu/~wc9c/KING/manual.html#WITHIN</a></p></li>
</ul>
</dd></dl>

</div>
</div>
</div>
</div>
<div class="section" id="module-cgr_gwas_qc.parsers.plink">
<span id="plink"></span><h2>Plink<a class="headerlink" href="#module-cgr_gwas_qc.parsers.plink" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt id="cgr_gwas_qc.parsers.plink.read_genome">
<code class="sig-prename descclassname">cgr_gwas_qc.parsers.plink.</code><code class="sig-name descname">read_genome</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">filename</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>os.PathLike<span class="p">[</span>str<span class="p">]</span><span class="p">, </span>pathlib.Path<span class="p">]</span></span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="reference internal" href="../_modules/cgr_gwas_qc/parsers/plink.html#read_genome"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cgr_gwas_qc.parsers.plink.read_genome" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse PLINK’s genome file format.</p>
<p>Each row of the genome file is a pairwise combinations of
samples/subjects. I am unsure of how plink assigns IID order. Here I sort
IDs alphanumerically which will make searching for pairs easier because
you can assume the order.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ID1</span><span class="p">,</span> <span class="n">ID2</span> <span class="o">=</span> <span class="n">sort</span><span class="p">([</span><span class="n">IID1</span><span class="p">,</span> <span class="n">IID2</span><span class="p">])</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><p>A (n x 5) table with the following columns</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 13%" />
<col style="width: 13%" />
<col style="width: 13%" />
<col style="width: 13%" />
<col style="width: 13%" />
<col style="width: 13%" />
<col style="width: 13%" />
<col style="width: 13%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>name</p></th>
<th class="head"><p>dtype</p></th>
<th class="head"><p>description</p></th>
<th class="head"></th>
<th class="head"></th>
<th class="head"></th>
<th class="head"></th>
<th class="head"></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>ID1</p></td>
<td><p>object</p></td>
<td><p>First Sample or Subject ID (alphanumerically)</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>ID2</p></td>
<td><p>object</p></td>
<td><p>Second Sample or Subject ID (alphanumerically)</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>RT</p></td>
<td><p>object</p></td>
<td><p>Relationship type inferred from .fam/.ped file {FS: Full Sib</p></td>
<td><p>HS</p></td>
<td><p>Half Sib</p></td>
<td><p>PO: Parent-Offspring</p></td>
<td><p>OT; Other}</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>EZ</p></td>
<td><p>object</p></td>
<td><p>IBD sharing expected value</p></td>
<td><p>based on just .fam/.ped relationship</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>Z0</p></td>
<td><p>float</p></td>
<td><p>P(IBD=0)</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>Z1</p></td>
<td><p>float</p></td>
<td><p>P(IBD=1)</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>Z2</p></td>
<td><p>float</p></td>
<td><p>P(IBD=2)</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>PI_HAT</p></td>
<td><p>float</p></td>
<td><p>Proportion IBD</p></td>
<td><p>i.e. P(IBD=2) + 0.5*P(IBD=1)</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>PHE</p></td>
<td><p>int</p></td>
<td><p>Pairwise phenotypic code (1</p></td>
<td><p>0</p></td>
<td><p>-1 = case-case</p></td>
<td><p>case-ctrl</p></td>
<td><p>and ctrl-ctrl pairs</p></td>
<td><p>respectively)</p></td>
</tr>
<tr class="row-odd"><td><p>DST</p></td>
<td><p>float</p></td>
<td><p>IBS distance</p></td>
<td><p>i.e. (IBS2 + 0.5*IBS1) / (IBS0 + IBS1 + IBS2)</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>PPC</p></td>
<td><p>float</p></td>
<td><p>IBS binomial test</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>RATIO</p></td>
<td><p>float</p></td>
<td><p>HETHET: IBS0 SNP ratio (expected value 2)</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>IBS0</p></td>
<td><p>int</p></td>
<td><p>Number of IBS 0 nonmissing variants</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>IBS1</p></td>
<td><p>int</p></td>
<td><p>Number of IBS 1 nonmissing variants</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>IBS2</p></td>
<td><p>int</p></td>
<td><p>Number of IBS 2 nonmissing variants</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>HOMHOM</p></td>
<td><p>float</p></td>
<td><p>Number of IBS 0 SNP pairs used in PPC test</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>HETHET</p></td>
<td><p>float</p></td>
<td><p>Number of IBS 2 het/het SNP pairs used in PPC test</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
<p class="rubric">References</p>
<ul class="simple">
<li><p><a class="reference external" href="https://www.cog-genomics.org/plink/1.9/ibd">https://www.cog-genomics.org/plink/1.9/ibd</a></p></li>
<li><p><a class="reference external" href="https://www.cog-genomics.org/plink/1.9/formats#genome">https://www.cog-genomics.org/plink/1.9/formats#genome</a></p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt id="cgr_gwas_qc.parsers.plink.read_het">
<code class="sig-prename descclassname">cgr_gwas_qc.parsers.plink.</code><code class="sig-name descname">read_het</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">filename</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>os.PathLike<span class="p">[</span>str<span class="p">]</span><span class="p">, </span>pathlib.Path<span class="p">]</span></span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="reference internal" href="../_modules/cgr_gwas_qc/parsers/plink.html#read_het"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cgr_gwas_qc.parsers.plink.read_het" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse PLINK’s het file format.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><p>A (n x 5) table with the following columns</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>name</p></th>
<th class="head"><p>dtype</p></th>
<th class="head"><p>description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>ID</strong> (<em>index</em>)</p></td>
<td><p>object</p></td>
<td><p>Sample or Subject ID</p></td>
</tr>
<tr class="row-odd"><td><p>O_HOM</p></td>
<td><p>int</p></td>
<td><p>Observed number of homozygotes</p></td>
</tr>
<tr class="row-even"><td><p>E_HOM</p></td>
<td><p>int</p></td>
<td><p>Expected number of homozygotes</p></td>
</tr>
<tr class="row-odd"><td><p>N_NM</p></td>
<td><p>int</p></td>
<td><p>Number of non-missing autosomal genotypes</p></td>
</tr>
<tr class="row-even"><td><p>F</p></td>
<td><p>float</p></td>
<td><p>Method-of-moments F coefficient estimate</p></td>
</tr>
</tbody>
</table>
</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
<p class="rubric">References</p>
<ul class="simple">
<li><p><a class="reference external" href="https://www.cog-genomics.org/plink/1.9/basic_stats#ibc">https://www.cog-genomics.org/plink/1.9/basic_stats#ibc</a></p></li>
<li><p><a class="reference external" href="https://www.cog-genomics.org/plink/1.9/formats#het">https://www.cog-genomics.org/plink/1.9/formats#het</a></p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt id="cgr_gwas_qc.parsers.plink.read_hwe">
<code class="sig-prename descclassname">cgr_gwas_qc.parsers.plink.</code><code class="sig-name descname">read_hwe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">filename</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>os.PathLike<span class="p">[</span>str<span class="p">]</span><span class="p">, </span>pathlib.Path<span class="p">]</span></span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="reference internal" href="../_modules/cgr_gwas_qc/parsers/plink.html#read_hwe"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cgr_gwas_qc.parsers.plink.read_hwe" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse PLINK’s hwe file format.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><p>A (# SNP x 9) table with the following columns</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 14%" />
<col style="width: 14%" />
<col style="width: 14%" />
<col style="width: 14%" />
<col style="width: 14%" />
<col style="width: 14%" />
<col style="width: 14%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>name</p></th>
<th class="head"><p>dtype</p></th>
<th class="head"><p>description</p></th>
<th class="head"></th>
<th class="head"></th>
<th class="head"></th>
<th class="head"></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>SNP</strong> (<em>index</em>)</p></td>
<td><p>string</p></td>
<td><p>SNP ID</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>CHR</p></td>
<td><p>string</p></td>
<td><p>Chromosome code</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>TEST</p></td>
<td><p>string</p></td>
<td><p>Type of test; one of {ALL</p></td>
<td><p>AFF</p></td>
<td><p>UNAFF</p></td>
<td><p>ALL(QT)</p></td>
<td><p>ALL(NP)}</p></td>
</tr>
<tr class="row-odd"><td><p>A1</p></td>
<td><p>string</p></td>
<td><p>Allele 1 (usually minor)</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>A2</p></td>
<td><p>string</p></td>
<td><p>Allele 2 (usually major)</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>GENO</p></td>
<td><p>string</p></td>
<td><p>’/’- separated genotype counts (A1 hom</p></td>
<td><p>het</p></td>
<td><p>A2 hom)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>O_HET</p></td>
<td><p>float</p></td>
<td><p>Observed heterozygote frequency</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>E_HET</p></td>
<td><p>float</p></td>
<td><p>Expected heterozygote frequency</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>P</p></td>
<td><p>float</p></td>
<td><p>Hardy-Weinberg equilibrium exact test p-value</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
<p class="rubric">References</p>
<ul class="simple">
<li><p><a class="reference external" href="https://www.cog-genomics.org/plink/1.9/basic_stats#hardy">https://www.cog-genomics.org/plink/1.9/basic_stats#hardy</a></p></li>
<li><p><a class="reference external" href="https://www.cog-genomics.org/plink/1.9/formats#hwe">https://www.cog-genomics.org/plink/1.9/formats#hwe</a></p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt id="cgr_gwas_qc.parsers.plink.read_imiss">
<code class="sig-prename descclassname">cgr_gwas_qc.parsers.plink.</code><code class="sig-name descname">read_imiss</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">filename</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>os.PathLike<span class="p">[</span>str<span class="p">]</span><span class="p">, </span>pathlib.Path<span class="p">]</span></span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="reference internal" href="../_modules/cgr_gwas_qc/parsers/plink.html#read_imiss"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cgr_gwas_qc.parsers.plink.read_imiss" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse PLINK’s imiss file format.</p>
<p>This reports sample level missing data.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><p>A (n x 5) table with the following columns</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>name</p></th>
<th class="head"><p>dtype</p></th>
<th class="head"><p>description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>ID</strong> (<em>index</em>)</p></td>
<td><p>object</p></td>
<td><p>Sample or Subject ID</p></td>
</tr>
<tr class="row-odd"><td><p>MISS_PHENO</p></td>
<td><p>str</p></td>
<td><p>[Y/N] if the phenotype is missing</p></td>
</tr>
<tr class="row-even"><td><p>N_MISS</p></td>
<td><p>int</p></td>
<td><p>The number of missing genotype calls not including obligatory missing or heterozygous haploids.</p></td>
</tr>
<tr class="row-odd"><td><p>N_GENO</p></td>
<td><p>int</p></td>
<td><p>Number of potentially valid calls</p></td>
</tr>
<tr class="row-even"><td><p>F_MISS</p></td>
<td><p>float</p></td>
<td><p>Missing call rate</p></td>
</tr>
</tbody>
</table>
</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
<p class="rubric">References</p>
<ul class="simple">
<li><p><a class="reference external" href="https://www.cog-genomics.org/plink/1.9/basic_stats#missing">https://www.cog-genomics.org/plink/1.9/basic_stats#missing</a></p></li>
<li><p><a class="reference external" href="https://www.cog-genomics.org/plink/1.9/formats#imiss">https://www.cog-genomics.org/plink/1.9/formats#imiss</a></p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt id="cgr_gwas_qc.parsers.plink.read_lmiss">
<code class="sig-prename descclassname">cgr_gwas_qc.parsers.plink.</code><code class="sig-name descname">read_lmiss</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">filename</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>os.PathLike<span class="p">[</span>str<span class="p">]</span><span class="p">, </span>pathlib.Path<span class="p">]</span></span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="reference internal" href="../_modules/cgr_gwas_qc/parsers/plink.html#read_lmiss"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cgr_gwas_qc.parsers.plink.read_lmiss" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse PLINK’s lmiss file format.</p>
<p>This reports snp level missing data.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><p>A (n x 5) table with the following columns</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>name</p></th>
<th class="head"><p>dtype</p></th>
<th class="head"><p>description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>CHR</p></td>
<td><p>str</p></td>
<td><p>Chromosome code</p></td>
</tr>
<tr class="row-odd"><td><p>SNP</p></td>
<td><p>str</p></td>
<td><p>Variant identifier</p></td>
</tr>
<tr class="row-even"><td><p>N_MISS</p></td>
<td><p>int</p></td>
<td><p>The number of missing genotype calls not including obligatory missing or heterozygous haploids.</p></td>
</tr>
<tr class="row-odd"><td><p>N_GENO</p></td>
<td><p>int</p></td>
<td><p>Number of potentially valid calls</p></td>
</tr>
<tr class="row-even"><td><p>F_MISS</p></td>
<td><p>float</p></td>
<td><p>Missing call rate</p></td>
</tr>
</tbody>
</table>
</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
<p class="rubric">References</p>
<ul class="simple">
<li><p><a class="reference external" href="https://www.cog-genomics.org/plink/1.9/basic_stats#missing">https://www.cog-genomics.org/plink/1.9/basic_stats#missing</a></p></li>
<li><p><a class="reference external" href="https://www.cog-genomics.org/plink/1.9/formats#lmiss">https://www.cog-genomics.org/plink/1.9/formats#lmiss</a></p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt id="cgr_gwas_qc.parsers.plink.read_sexcheck">
<code class="sig-prename descclassname">cgr_gwas_qc.parsers.plink.</code><code class="sig-name descname">read_sexcheck</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">filename</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>os.PathLike<span class="p">[</span>str<span class="p">]</span><span class="p">, </span>pathlib.Path<span class="p">]</span></span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="reference internal" href="../_modules/cgr_gwas_qc/parsers/plink.html#read_sexcheck"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cgr_gwas_qc.parsers.plink.read_sexcheck" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse PLINK’s sexcheck file format.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><p>A (n x 5) table with the following columns</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>name</p></th>
<th class="head"><p>dtype</p></th>
<th class="head"><p>description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>ID</strong> (<em>index</em>)</p></td>
<td><p>object</p></td>
<td><p>Sample or Subject ID</p></td>
</tr>
<tr class="row-odd"><td><p>PEDSEX</p></td>
<td><p>int</p></td>
<td><p>Sex code in input file</p></td>
</tr>
<tr class="row-even"><td><p>SNPSEX</p></td>
<td><p>int</p></td>
<td><p>Imputed sex code (1 = male; 2 = female; 0 = unknown)</p></td>
</tr>
<tr class="row-odd"><td><p>STATUS</p></td>
<td><p>str</p></td>
<td><p>OK if PEDSEX and SNPSEX match and are nonzero PROBLEM otherwise</p></td>
</tr>
<tr class="row-even"><td><p>F</p></td>
<td><p>int</p></td>
<td><p>Inbreeding coefficient considering only X chromosome.</p></td>
</tr>
</tbody>
</table>
</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
<p class="rubric">References</p>
<ul class="simple">
<li><p><a class="reference external" href="https://www.cog-genomics.org/plink/1.9/basic_stats#check_sex">https://www.cog-genomics.org/plink/1.9/basic_stats#check_sex</a></p></li>
<li><p><a class="reference external" href="https://www.cog-genomics.org/plink/1.9/formats#sexcheck">https://www.cog-genomics.org/plink/1.9/formats#sexcheck</a></p></li>
</ul>
</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="workflow.html" class="btn btn-neutral float-right" title="Workflow Scripts" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="config.html" class="btn btn-neutral float-left" title="Configuration" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Leidos Biomedical Research, Inc.

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>